File: quicksort/quicksort.go
package main

import (
	"sync"
	"time"
	"os"
	"math/rand"
	"strconv"
	"fmt"
)

func QuickSortSequential(a *[]int, low, hi int) {
	if hi < low {
		return
	}

	p := partition(a, low, hi)
	QuickSortSequential(a, low, p-1)
	QuickSortSequential(a, p+1, hi)
}

func QuickSortConcurrent(a *[]int, low, hi, depth int) {
	if hi < low {
		return
	}

	p := partition(a, low, hi)

	if depth > 0 {
		wg := sync.WaitGroup{}
		wg.Add(2)
		go func(){
			QuickSortConcurrent(a, low, p-1, depth-1)
			wg.Done()
		}()
		go func(){
			QuickSortConcurrent(a, p+1, hi, depth-1)
			wg.Done()
		}()
		wg.Wait()
	}else{
		QuickSortSequential(a,low, p-1)
		QuickSortSequential(a, p+1, hi)
	}	
}

func partition(a *[]int, low, hi int) int {
	pivot := (*a)[hi]
	i := low - 1

	for j := low; j < hi; j++{
		if (*a)[j] < pivot {
			i++
			swap(a,i,j)
		}
	}

	if (*a)[hi] < (*a)[i + 1] {
		swap(a,hi,i+1)
	}

	return i + 1
}

func swap(a *[]int, i , j int) {
	(*a)[i], (*a)[j] = (*a)[j], (*a)[i]
}

func main(){
	start := time.Now()
	n,_ := strconv.Atoi(os.Args[1])
	
	rand.Seed(time.Now().UTC().UnixNano())
	list := rand.Perm(n)
	fmt.Println(time.Since(start).Seconds())
	start = time.Now()
	if len(os.Args) > 2 {
		depth,_ := strconv.Atoi(os.Args[2])
		QuickSortConcurrent(&list,0, len(list)-1, depth)
	}else {
		QuickSortSequential(&list,0, len(list)-1)
	}
	fmt.Println(time.Since(start).Seconds())
}

File: quicksort/quicksort.cpp
//source: https://github.com/pHag/id1217/blob/master/hw2/ex2/quickSort-openmp.c

#include <omp.h>
#include <time.h>
#include <sys/time.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

void quickSort(int *inputArray, int size, int depth);
void swap(int *inputArray, int leftIndex, int rightIndex);

double start_time, end_time; /* start and end times */
long size; /* array size */
/* ---------------------------------------------------------------------------- */
/* read command line, initialize, and create threads */
int main(int argc, char *argv[]) {
	int i, depth, numWorkers;

	size = atoi(argv[1]);

	depth = (argc > 2)? atoi(argv[2]) : 0;
	numWorkers =(depth > 0)? pow(2,depth)-1 : 1;
	omp_set_num_threads(numWorkers); /* Set number of threads */

	int *inputArray; /* testDataArray */
	inputArray = (int*)malloc(sizeof(int) * size); /* Allocate in memory instead */
	srand(time(NULL));
	/* Create testData array */
	for (i = 0; i < size; i++) {
		inputArray[i] = rand()%99999;
	}

	struct timeval t1, t2;
	double elapsedTime;
    gettimeofday(&t1, NULL);
	/* Call the quickSort function to sort the list */
	#pragma omp parallel
	{
		/* We only want our master thread to be executed once, thus we use the singel construct here.
			nowait is used becuse we have no need for synchronization at the end of the region */
		#pragma omp single nowait
		{
			quickSort(inputArray, size, depth);
		}
	}
	gettimeofday(&t2, NULL);
	elapsedTime = (t2.tv_sec - t1.tv_sec) + (t2.tv_usec - t1.tv_usec)/1000000.0;
	printf("%f\n",elapsedTime);

	free(inputArray);
	return 0;
}

void quickSort(int *inputArray, int size, int depth){
	int pivot, leftIndex, rightIndex;
	/* End of reccursion */
	if (size <= 1) { return; }
	/* Set pivot */
	pivot = inputArray[size/2];
	for(leftIndex = 0, rightIndex = size -1;; leftIndex++, rightIndex--) {
		while(inputArray[leftIndex] < pivot){
			leftIndex++;
		}
		while(pivot < inputArray[rightIndex]){
			rightIndex--;
		}
		if(rightIndex <= leftIndex){
			break;
		}
		swap(inputArray, leftIndex, rightIndex);    
	}
	if(depth == 0){
		quickSort(inputArray, leftIndex, 0);
		quickSort(inputArray + rightIndex + 1, size - rightIndex -1, 0);
		return;
	}
	
	#pragma omp task
	{
		quickSort(inputArray, leftIndex, depth-1); /* Sort lower */
	}
	#pragma omp task
	{
		quickSort(inputArray + rightIndex + 1, size - rightIndex -1, depth-1); /* Sort upper */
	}
}
/* ---------------------------------------------------------------------------- */
/* Swaps two elements */
void swap(int *inputArray, int leftIndex, int rightIndex){
	int temp;
	temp = inputArray[leftIndex];
	inputArray[leftIndex] = inputArray[rightIndex];
	inputArray[rightIndex] = temp;
}
